### PR TITLE
Force inherit BDVD/PS3_GAME across game updates

### LINKED ISSUE #13436: Demon's Souls [NPJA00102] can't boot when applied 1.04 update.
### Quick summary

Demon's Souls [NPJA00102] can't boot when applied update 1.04.
Update pkg is same as disk version (BCJS30022).

### Details

In PS3, NPJA00102 can be aplied update 1.04, and ps3 create directory named "NPJA00102" and "BCJS30032".
RPCS3 install update pkg same as real PS3, but it cause fatal error on boot.

### Attach a log file

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/10810731/RPCS3.log.gz)
スクリーンショット 2023-02-23 151411

### Attach capture files for visual issues

_No response_

### System configuration

_No response_

### Other details

_No response_

### PR DESCRIPTION
**Pull request by elad335**
Fixes #13436

### CODE REVIEW

File: rpcs3/Emu/System.cpp [hunk 1]
```diff
@@ -1746,6 +1751,11 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 				sys_log.success("Updates found at /dev_hdd0/game/%s/", m_title_id);
 				m_path = hdd0_boot;
 
+				if (m_title_id == "DG")
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:1754]: What's this supposed to do?

File: rpcs3/Emu/System.cpp [hunk 2]
```diff
@@ -1587,9 +1588,9 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 				vfs::mount("/dev_hdd0/game/" + m_title_id, game_dir + '/');
 			}
 		}
-		else if (m_cat == "DG" && from_hdd0_game && disc.empty())
+		else if (!inherited_ps3_game.empty() || (m_cat == "DG" && disc.empty()))
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:1591]: Why remove the condition? We don't want people to just paste their games anywhere.

File: rpcs3/Emu/System.cpp [hunk 3]
```diff
@@ -1598,8 +1599,12 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 				return game_boot_result::invalid_bdvd_folder;
 			}
 
+			// TODO: Check this
 			vfs::mount("/dev_bdvd", bdvd_dir);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:1603]: We already checked this. It's what was needed for wkc or whatever

Comment (elad335) [reply rpcs3/Emu/System.cpp:1603]: I mean what is the actual path used.

File: rpcs3/Emu/System.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/Emu/System.cpp:1591]: It's HW correct, PS3_GAME is actually a virtual directory. LV2 disaaembly has even a specific check for it.

File: rpcs3/Emu/System.cpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1603]: It doesn't matter, as there is no disc, so you can't read or write any files to the folder

Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1610]: And the virtualization is oblivious to the game

Comment (elad335) [reply rpcs3/Emu/System.cpp:1610]: It matters for timestamps obtainining and other sys_fs operations.

File: rpcs3/Emu/System.cpp (hunk 2 above)
Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1598]: I mean why remove from_hdd0_game check.  Where else would it be

Comment (elad335) [reply rpcs3/Emu/System.cpp:1598]: Anywhere. Like in /host_root. I discovered it by accident when testing.
host_root is the mount point of the host filesystem, the one which boots external software from remote PC to debug console for example.

Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1598]: This also makes the else path obsolete.
I don't think we should remove it.
There is no reason to support installed games outside of hdd0

Comment (elad335) [reply rpcs3/Emu/System.cpp:1598]: What if I make a hw test and put it outside HDD0? why should the result differ from PS3?

Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1598]: Because we are dealing with smoothbrains in the discord help section.

Comment (Megamouse) [reply rpcs3/Emu/System.cpp:1598]: You're literally the only one ever trying this.
Why would you expose such a potential desaster to everyone.

Comment (elad335) [reply rpcs3/Emu/System.cpp]: Reverted.

File: rpcs3/Emu/System.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/Emu/System.cpp:1616]: Clarified.

File: rpcs3/Emu/System.h [hunk 4]
```diff
@@ -216,6 +216,7 @@ class Emulator final
 	std::vector<u8> data;
 	std::vector<u128> klic;
 	std::string disc;
+	std::string inherited_ps3_game;
```
Comment (Megamouse) [inline rpcs3/Emu/System.h:219]: Why not call it
m_inherited_ps3_game
And put it to the other private members?

It seems out of place here.

File: rpcs3/Emu/System.cpp [hunk 5]
```diff
@@ -918,7 +918,14 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
...
 				if (std::string game_path = m_games_config.get_path(m_title_id); !game_path.empty())
 				{
-					disc = std::move(game_path);
+					if (fs::is_file(game_path + "/PS3_DISC.SFB"))
+					{
+						disc = std::move(game_path);
+					}
+					else
+					{
+						inherited_ps3_game = std::move(game_path);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:927]: I'm missing the .clear() call somewhere

Comment (Megamouse) [reply rpcs3/Emu/System.cpp]: Actually, it seems like there is no reason for it to be a member anymore I think

Comment (elad335) [reply rpcs3/Emu/System.cpp]: Yeah I made it a local.

File: rpcs3/Emu/System.cpp [hunk 6]
```diff
@@ -2561,9 +2588,15 @@ void Emulator::Kill(bool allow_autoexit, bool savestate)
 		m_config_mode = cfg_mode::custom;
 		read_used_savestate_versions();
 		m_savestate_extension_flags1 = {};
+
+		// Enable logging
+		rpcs3::utils::configure_logs(true);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:2593]: What about the other return?

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 7]
```diff
@@ -133,14 +133,26 @@ void fmt_class_string<typename ppu_thread::call_history_t>::format(std::string&
...
 	{
-		const u32 pc = history.data[idx % ppu_thread::call_history_max_size];
+		const u32 pc = history.data[idx % history.data.size()];
 		dis_asm.disasm(pc);
 		fmt::append(out, "\n(%u) 0x%08x: %s", count, pc, dis_asm.last_opcode);
 	}
 }
 
+template <>
+void fmt_class_string<typename ppu_thread::syscall_history_t>::format(std::string& out, u64 arg)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/PPUThread.cpp:145]: This stuff is unrelated to the topic and should be its own PR

File: rpcs3/Emu/System.cpp (hunk 6 above)
Comment (elad335) [reply rpcs3/Emu/System.cpp:2593]: The other one is an error path, it should do nothing. configure_logs should have already been called there.

File: rpcs3/Emu/System.cpp [hunk 8]
```diff
@@ -1598,9 +1614,23 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 				return game_boot_result::invalid_bdvd_folder;
 			}
 
+			// TODO: Verify timestamps and error codes with  sys_fs
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:1617]:
```suggestion
			// TODO: Verify timestamps and error codes with sys_fs
```
